### PR TITLE
adaptive concurrency: Add buffer to minRTT measurements

### PR DESCRIPTION
**Pull request by tonya11en**
This patch modifies the concurrency limit calculation to increase the measured minRTT by 25% in the formula. This allows for configuration of a threshold that must be exceeded before the concurrency limit begins to decrease.

In addition, the gradient value is clamped between 0.5 and 2. The option to configure it via proto or runtime has been removed.

**Motivation**
Analysis of adaptive concurrency behavior in the Lyft infrastructure has shown that the gradient controller tends to overreact to normal variance in the latencies. If an upstream's p50 latency hovers around 20ms, a measurement of 25ms should not cause the concurrency limit to drop 20%. In theory, the observed latencies should not deviate from the minRTT by any appreciable amount in the steady-state, but in reality there is noise introduced to the measurements from environmental factors. This buffer value is necessary for well-behaved adaptive concurrency control and should improve the overall stability of the concurrency limits.

### CODE REVIEW

File: docs/root/configuration/http/http_filters/adaptive_concurrency_filter.rst [hunk 1]
```diff
@@ -64,9 +64,11 @@ The gradient is calculated using summarized sampled request latencies (sampleRTT
 
 .. math::
 
-    gradient = \frac{minRTT}{sampleRTT}
+    gradient = \frac{minRTT + buffer}{sampleRTT}
```
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/adaptive_concurrency_filter.rst]: Can you add more of an explanation in the main docs on how the buffer is calculated and what it is for? Also a ref link to the configuration field would be useful.

File: source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc [hunk 2]
```diff
@@ -148,10 +147,12 @@ std::chrono::microseconds GradientController::processLatencySamplesAndClear() {
 }
 
 uint32_t GradientController::calculateNewLimit() {
-  // Calculate the gradient value, ensuring it remains below the configured maximum.
   ASSERT(sample_rtt_.count() > 0);
-  const double raw_gradient = static_cast<double>(min_rtt_.count()) / sample_rtt_.count();
-  const double gradient = std::min(config_.maxGradient(), raw_gradient);
+
+  // Calculate the gradient value, ensuring it remains below the configured maximum.
+  auto buffered_min_rtt = min_rtt_.count() + min_rtt_.count() * config_.minRTTBufferPercent();
```
Comment (mattklein123) [inline source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc]: nit: const

File: source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc [hunk 3]
```diff
@@ -37,13 +37,12 @@ GradientControllerConfig::GradientControllerConfig(
...
-                                                    max_gradient, 2.0)),
       sample_aggregate_percentile_(
           PROTOBUF_PERCENT_TO_DOUBLE_OR_DEFAULT(proto_config, sample_aggregate_percentile, 50)),
-      min_concurrency_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config.min_rtt_calc_params(),
-                                                       min_concurrency, 3)) {}
-
+      min_concurrency_(
+          PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config.min_rtt_calc_params(), min_concurrency, 3)),
+      min_rtt_buffer_pct_(
+          PROTOBUF_PERCENT_TO_DOUBLE_OR_DEFAULT(proto_config.min_rtt_calc_params(), buffer, 25)) {}
```
Comment (mattklein123) [inline source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc]: In the docs you say the default is 10. Should it be 10 here or 25 in the docs?

File: source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc [hunk 4]
```diff
@@ -148,10 +147,12 @@ std::chrono::microseconds GradientController::processLatencySamplesAndClear() {
...
 uint32_t GradientController::calculateNewLimit() {
-  // Calculate the gradient value, ensuring it remains below the configured maximum.
   ASSERT(sample_rtt_.count() > 0);
-  const double raw_gradient = static_cast<double>(min_rtt_.count()) / sample_rtt_.count();
-  const double gradient = std::min(config_.maxGradient(), raw_gradient);
+
+  // Calculate the gradient value, ensuring it remains below the configured maximum.
+  const auto buffered_min_rtt = min_rtt_.count() + min_rtt_.count() * config_.minRTTBufferPercent();
+  const double raw_gradient = static_cast<double>(buffered_min_rtt) / sample_rtt_.count();
+  const double gradient = std::max<double>(0.5, std::min<double>(2.0, raw_gradient));
```
Comment (mattklein123) [inline source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc]: How did you determine 0.5 as the lower bound for the clamp? Can you discuss this clamping of the gradient in the docs a bit? (Also per our offline convo I really do think a simulation utility would be very useful for evaluating algorithm changes)

File: source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc (hunk 3 above)
Comment (tonya11en) [reply source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc]: Ah, oversight on my part. 25% everywhere.

File: source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc (hunk 4 above)
Comment (tonya11en) [reply source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc]: I put some thought into the simulation and keep coming back to something that looks like the bufferbloater tests, but implemented in the Envoy codebase to test the concurrency controllers. What complicates things here compared to the load balancer simulation is that the LB simulation is validating a certain request distribution after many requests, but an adaptive concurrency simulation requires us to validate its temporal behavior and how that would affect upstream latency.

This is achievable and I'll chip away at it here in the background. The next algorithm change will be prefaced with a simulator MVP.

Comment (mattklein123) [reply source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc]: > Can you discuss this clamping of the gradient in the docs a bit?

Ping on this but feel free to do it as part of your next change.

**Pull request merged**